Add support for per-shape physics materials

tancop opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/7401.

This PR adds:
- Methods `shape_set_friction`,  `shape_get_friction`,  `shape_set_bounce`,  `shape_get_bounce` to `PhysicsServer3D`
- Property `physics_material` to `Shape3D`

This is implemented for both Godot and Jolt Physics (using a custom Jolt physics material like [one of the official samples](https://github.com/jrouwe/JoltPhysics/blob/master/Samples/Tests/General/FrictionPerTriangleTest.cpp)).

Example project that uses two different materials on one body: [per-shape-materials.zip](https://github.com/user-attachments/files/18373092/per-shape-materials.zip)

tancop commented:
Fixed the xml method order and added support for concave shapes.

tancop commented:
Added support for changing materials at runtime. I had to add a new custom shape type `JoltCustomMaterialOverrideShape`, add a new virtual method on `JoltShape3D` and change some others to return a mutable `JPH::Ref<JPH::Shape>` instead of a `JPH::ShapeRefC`.

tancop commented:
@AThousandShips I guess this is ready for review. If you see something wrong or confusing let me know

jrouwe commented on modules/jolt_physics/shapes/jolt_concave_polygon_shape_3d.cpp:
```diff
@@ -29,13 +29,25 @@
+		return;
+	}
+	jolt_ref_mutex.lock();
+	JoltCustomMaterialOverrideShape *ref = static_cast<JoltCustomMaterialOverrideShape *>(jolt_ref.GetPtr());
```
I don't think you need `JoltCustomMaterialOverrideShape`. You can use:

```
PhysicsMaterialRefC array[] = { mat };
MeshShape::RestoreMaterialState(array, 1);
```

instead.

jrouwe commented on modules/jolt_physics/shapes/jolt_height_map_shape_3d.cpp:
```diff
@@ -49,7 +50,30 @@ bool _is_triangle_hole(const JPH::VertexList &p_vertices, int p_index0, int p_in
+	jolt_ref_mutex.lock();
+
+	// The shape we built is always a `JoltCustomMaterialOverrideShape` with different inner shapes.
+	JoltCustomMaterialOverrideShape *ref = static_cast<JoltCustomMaterialOverrideShape *>(jolt_ref.GetPtr());
```
Same here.

jrouwe commented on modules/jolt_physics/shapes/jolt_world_boundary_shape_3d.cpp:
```diff
@@ -29,22 +29,34 @@
+		return;
+	}
+	jolt_ref_mutex.lock();
+	JoltCustomMaterialOverrideShape *ref = static_cast<JoltCustomMaterialOverrideShape *>(jolt_ref.GetPtr());
```
Same here (although I'll add the missing accessors for the material to this shape).

jrouwe commented on modules/jolt_physics/shapes/jolt_custom_decorated_shape.h:
```diff
@@ -55,6 +55,8 @@ class JoltCustomDecoratedShape : public JPH::DecoratedShape {
 
 	virtual JPH::MassProperties GetMassProperties() const override { return mInnerShape->GetMassProperties(); }
 
+	virtual const JPH::PhysicsMaterial *GetMaterial(const JPH::SubShapeID &p_sub_shape_id) const override { return mInnerShape->GetMaterial(p_sub_shape_id); }
```
This is the exact same implementation as JPH::DecoratedShape::GetMaterial, I think this can be left out.

tancop commented:
@jrouwe thanks! I saw `RestoreMaterialState` in the docs but thought it was just for deserializing from binary state. You saved 200 lines and a custom shape slot

jrouwe commented on modules/jolt_physics/shapes/jolt_custom_shape_type.h:
```diff
@@ -39,6 +39,7 @@ namespace JoltCustomShapeSubType {
 
 constexpr JPH::EShapeSubType OVERRIDE_USER_DATA = JPH::EShapeSubType::User1;
 constexpr JPH::EShapeSubType DOUBLE_SIDED = JPH::EShapeSubType::User2;
+constexpr JPH::EShapeSubType OVERRIDE_MATERIAL = JPH::EShapeSubType::User3;
```
This can be removed now

jrouwe commented on modules/jolt_physics/objects/jolt_shaped_object_3d.cpp:
```diff
@@ -167,7 +167,14 @@ void JoltShapedObject3D::_update_shape() {
+
+	if (uses_shape_materials()) {
+		// This body/area uses per-shape physics materials, manifold reduction is not safe (can merge shapes with different materials).
+		iface.SetUseManifoldReduction(jolt_id, false);
```
You may want to consider turning manifold reduction ON again if the shape no longer uses shape materials (and if `!reports_contacts()`). This saves some CPU cycles during collision detection.

tancop commented:
Renamed `JoltShape3D::_set_material` to `_update_material`, it only changes the generated Jolt material not the shape itself

tancop commented:
Rebased on latest `master`

aaronfranke commented:
Interesting that this is on the Shape3D resource, I was imagining it on the CollisionShape3D node. That's how it's done in the glTF physics. But anyway, I don't have any specific reason to dislike the PR as-is.

tancop commented:
> Interesting that this is on the Shape3D resource, I was imagining it on the CollisionShape3D node. That's how it's done in the glTF physics extensions. But anyway, I don't have any specific reason to dislike the PR as-is, aside from needing to write some more complex importer code.

I can change it if it's better that way. I just think putting it on `Shape3D` fits better when margin and solver bias are already there, `CollisionShape3D` is more like a container for shapes. This comes down to if we want to share material between nodes with the same shape or not.

aaronfranke commented:
Here's what the data looks like on the glTF side on a node.

* OMI: https://github.com/aaronfranke/gltf-extensions/blob/OMI_materials_blend/extensions/2.0/OMI_physics_body/schema/node.OMI_physics_body.collider.schema.json
* Microsoft: https://github.com/eoineoineoin/glTF_Physics/blob/master/extensions/2.0/Khronos/KHR_physics_rigid_bodies/schema/node.KHR_physics_rigid_bodies.collider.schema.json

aaronfranke commented on modules/jolt_physics/shapes/jolt_physics_material.h:
```diff
@@ -0,0 +1,48 @@
+			friction(p_friction), bounce(p_bounce) {}
+
+	real_t friction;
+	real_t bounce;
```
Variables should go above the constructor.

Also, does Jolt not have friction/bounce combine modes in its physics materials? glTF has ["average", "minimum", "maximum", "multiply"], Godot has `bool absorbent` and `bool rough`.

tancop replied:
Jolt physics materials are just a container for custom per-shape data, there are no combine modes only equations that combine the two values. Jolt uses `friction = sqrt(friction1 * friction2)` and `bounce = max(bounce1, bounce2)` by default so I copied them

jrouwe replied:
Godot Jolt does override this behavior:

https://github.com/godotengine/godot/blob/d2ada64a03d2abdb97cafe8f10623db8a2ce1d4c/modules/jolt_physics/spaces/jolt_space_3d.cpp#L124-L130

tancop replied:
Don't know why I never noticed that. Changed the combine formulas to match Godot Jolt for now, not sure if it's more or less accurate than Jolt's defaults

jrouwe replied:
imo it's quite an arbitrary choice how you combine restitution / friction. The reason why it differs from the default is because godot physics used these formulas and mihe wanted to keep the differences between the two implementations to a minimum.

jrouwe replied:
B.t.w. @mihe I would appreciate if you could take a look at this one. It is quite a substantial change and I can only judge how it interacts with Jolt, not how it should interact with godot.

mihe replied:
Yeah, I'm planning to (and I believe am expected to) take a closer look at this. Given that 4.4-beta1 (and its feature freeze) is just around the corner I would imagine that something as big as this is better suited for after 4.4-stable though.

aaronfranke reviewed (commented):
@tancop Thanks! I don't have an opinion on exposing that function.

mihe commented on servers/physics_server_3d.h:
```diff
@@ -277,6 +277,12 @@ class PhysicsServer3D : public Object {
 	virtual void shape_set_margin(RID p_shape, real_t p_margin) = 0;
 	virtual real_t shape_get_margin(RID p_shape) const = 0;
 
+	virtual real_t shape_get_friction(RID p_shape) const = 0;
+	virtual void shape_set_friction(RID p_shape, real_t p_friction) = 0;
+
+	virtual real_t shape_get_bounce(RID p_shape) const = 0;
+	virtual void shape_set_bounce(RID p_shape, real_t p_bounce) = 0;
```
I'll do a more in-depth review of the code/design at some later point, but I think this feature could be a good candidate for something like the server-specific properties discussed [here](https://github.com/godotengine/godot-proposals/issues/10438#issuecomment-2306922209), rather than tacking on stuff to the shared `PhysicsServer3D` interface that can't (?) be implemented by both physics engines (and perhaps also current/future physics extensions).

Merging this as-is would (in my opinion) set the somewhat unfortunate precedent of adding "Jolt-only" stuff to the shared `PhysicsServer3D` interface. If anything, we should aim to deprecate and move things away from the shared interface, to instead be provided per-server, for things that aren't deemed widely applicable to all/most other physics engines out there.

From just a quick look, I am seeing this supported by PhysX, Havok (judging by the Unity Physics and Babylon.js APIs), Rapier, bepuphysics, Newton and ODE, but not Bullet, so perhaps this concept is leaning more towards widely applicable either way, but with the Jolt module currently being considered experimental, and Godot Physics being the default 3D physics engine for the foreseeable future (theoretically forever) it would still seem unfortunate to expose this through `PhysicsServer3D`.

Exposing this through `PhysicsServer3D` also raises the question why this isn't exposed in `PhysicsServer2D`, even if just as an empty implementation.

With all that said, we obviously don't have the above mentioned per-server property functionality at the moment, which puts this PR in a weird limbo, but I'm hoping we can have it added in some basic form shortly after 4.4-stable is out.

aaronfranke replied:
Wouldn't the best option be to implement per-shape materials in Godot Physics, to match most physics engines? I don't know how much of a lift that would be.

mihe replied:
Sure, if that's added to this PR, then it's a different story, but I also don't know what's involved with adding per-shape materials to Godot Physics, nor what the consequences of adding that might be.

From just a quick look, it would seem to me like the shape information is available in the places where bounce and friction are being combined (`GodotBodyPair*D::pre_solve` and `GodotBodyPair*D::solve` respectively) but more than that I can't really say.

tancop replied:
I'll try to add Godot Physics support. Started out Jolt exclusive because that's what I use for all my projects

tancop commented:
Added Godot Physics support. Tried it on a test project and it works as expected, not as smooth as Jolt but good enough for me.

Also should we switch to Jolt's friction/bounce combine formulas for all physics engines? It feels more natural when the collision takes the average of both materials instead of just the lower friction one. Could do a follow up pr for that

mihe commented:
> Added Godot Physics support. Tried it on a test project and it works as expected, not as smooth as Jolt but good enough for me.

This should ideally be done for 2D as well in that case.

I'm hesitant about reviewing the Godot Physics part myself though. It's a simple enough of a change, but I have no idea what knock-on effects it might have, or what else might need to be changed to work with this. Maybe @rburing has some idea?

> Also should we switch to Jolt's friction/bounce combine formulas for all physics engines? It feels more natural when the collision takes the average of both materials instead of just the lower friction one. Could do a follow up pr for that

That seems like a significant breaking change, one which I don't see a strong motivation for.

tancop commented:
@mihe added 2D support and GDExtension bindings

Calinou reviewed (commented):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/b607110ad211778a6426a00ca9c8cd99c0eb383c), it works as expected. This will be handy for surface-dependent footsteps without needing to multiply StaticBodies per surface type :slightly_smiling_face: 

However, I got a crash when duplicating an empty CollisionShape2D in the editor. I could reproduce this several times as long as it has no shape assigned when it's being duplicated:

```
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.4.beta.custom_build (5d0cd6f635fd9663123efc1d15bbd913cb8e50df)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib64/libc.so.6(+0x1a050) [0x7fc9d4227050] (??:0)
[2] bin/godot.linuxbsd.editor.x86_64() [0x835b7e9] (/home/hugo/Documents/Git/godotengine/godot/./scene/2d/physics/collision_shape_2d.cpp:251 (discriminator 1))
[3] bin/godot.linuxbsd.editor.x86_64() [0x80eb7c1] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/binder_common.h:315 (discriminator 2))
[4] bin/godot.linuxbsd.editor.x86_64() [0x80e316f] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/binder_common.h:463)
[5] bin/godot.linuxbsd.editor.x86_64() [0x80d28a4] (/home/hugo/Documents/Git/godotengine/godot/./core/object/method_bind.h:345 (discriminator 1))
[6] bin/godot.linuxbsd.editor.x86_64() [0xac87ffc] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/variant.h:873)
[7] bin/godot.linuxbsd.editor.x86_64() [0xac9552b] (/home/hugo/Documents/Git/godotengine/godot/./core/object/object.cpp:284 (discriminator 1))
[8] bin/godot.linuxbsd.editor.x86_64() [0x744c25f] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/variant.h:873)
[9] bin/godot.linuxbsd.editor.x86_64() [0x744b5ee] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/node.cpp:2916)
[10] bin/godot.linuxbsd.editor.x86_64() [0x744b53c] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/node.cpp:2904 (discriminator 1))
[11] bin/godot.linuxbsd.editor.x86_64() [0x6447408] (/home/hugo/Documents/Git/godotengine/godot/./editor/scene_tree_dock.cpp:914)
[12] bin/godot.linuxbsd.editor.x86_64() [0x643ece1] (/home/hugo/Documents/Git/godotengine/godot/./editor/scene_tree_dock.cpp:196)
[13] bin/godot.linuxbsd.editor.x86_64() [0x7451b49] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/node.cpp:3479)
[14] bin/godot.linuxbsd.editor.x86_64() [0x750e641] (/home/hugo/Documents/Git/godotengine/godot/./core/templates/vector.h:230)
[15] bin/godot.linuxbsd.editor.x86_64() [0x753b219] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/viewport.cpp:3331)
[16] bin/godot.linuxbsd.editor.x86_64() [0x753ac99] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/viewport.cpp:3292)
[17] bin/godot.linuxbsd.editor.x86_64() [0x756b77b] (/home/hugo/Documents/Git/godotengine/godot/./scene/main/window.cpp:1703)
[18] bin/godot.linuxbsd.editor.x86_64() [0x7621770] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/binder_common.h:315 (discriminator 2))
[19] bin/godot.linuxbsd.editor.x86_64() [0x760efc2] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/binder_common.h:430)
[20] bin/godot.linuxbsd.editor.x86_64() [0x75f6924] (/home/hugo/Documents/Git/godotengine/godot/./core/object/callable_method_pointer.h:109)
[21] bin/godot.linuxbsd.editor.x86_64() [0xa72d2fb] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/callable.cpp:57)
[22] bin/godot.linuxbsd.editor.x86_64() [0x340d1ff] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/variant.h:952)
[23] bin/godot.linuxbsd.editor.x86_64() [0x33f2f91] (/home/hugo/Documents/Git/godotengine/godot/./core/variant/variant.h:873)
[24] bin/godot.linuxbsd.editor.x86_64() [0x33f2ca2] (/home/hugo/Documents/Git/godotengine/godot/platform/linuxbsd/x11/display_server_x11.cpp:4203)
[25] bin/godot.linuxbsd.editor.x86_64() [0xa696f4e] (/home/hugo/Documents/Git/godotengine/godot/./core/input/input.cpp:905)
[26] bin/godot.linuxbsd.editor.x86_64() [0xa698516] (/home/hugo/Documents/Git/godotengine/godot/./core/input/input.cpp:1186)
[27] bin/godot.linuxbsd.editor.x86_64() [0x33f8f56] (/home/hugo/Documents/Git/godotengine/godot/./core/templates/local_vector.h:380)
[28] bin/godot.linuxbsd.editor.x86_64() [0x33c4fcc] (/home/hugo/Documents/Git/godotengine/godot/platform/linuxbsd/os_linuxbsd.cpp:960)
[29] bin/godot.linuxbsd.editor.x86_64() [0x33ba371] (/home/hugo/Documents/Git/godotengine/godot/platform/linuxbsd/godot_linuxbsd.cpp:85)
[30] /lib64/libc.so.6(+0x3248) [0x7fc9d4210248] (??:0)
[31] /lib64/libc.so.6(__libc_start_main+0x8b) [0x7fc9d421030b] (??:0)
[32] bin/godot.linuxbsd.editor.x86_64() [0x33ba165] (??:?)
-- END OF BACKTRACE --
```

This crash also occurs with CollisionShape3D in the same conditions.

tancop commented:
@Calinou fixed it with a simple null check thanks for the heads up

Calinou reviewed (commented):
The crash is fixed now.

Looking again at the PhysicsServer API, I'm not sure about the usage of `NAN` as a "no override" value. We tend to avoid this kind of practice in Godot as it's known to be difficult to work with, especially in GDExtension bindings that may deal with NaNs differently. Likewise, setting `absorbent` and `rough` based on the value being negative is unintuitive. In terms of discoverability, it's better to have separate methods that show up in autocompletion.

To me, it seems more idiomatic to provide separate `shape_set_bounce_override(RID shape, bool enable)` and `shape_set_friction_override(RID shape, bool enable)` methods that accept a boolean, or even turn it into a single method `shape_set_friction_overide(RID shape, bool enable, float friction = 0.0)` (with the optional `friction` being ignored if `enable` is `false`).

`rough` and `absorbent` would be configured using similar override methods such as `shape_set_absorbent_override(RID shape, bool enable)`.

mihe commented:
> Likewise, setting `absorbent` and `rough` based on the value being negative is unintuitive. In terms of discoverability, it's better to have separate methods that show up in autocompletion.

Note that this is already the case (albeit undocumented) for the body-wide friction and bounce, where `rough` and `absorbent` are represented as negative values, so this is not new to this PR. I would personally prefer to see that the two are generally consistent in how they're set, even if the body-wide values are technically set through `body_set_param`, and not some `body_set_friction`.

Formally adding `rough` and `absorbent` to the physics server API would also further cement what is arguably a bit of a quirk in Godot, and might complicate efforts such as godotengine/godot-proposals#11715 that aims to add different ways of combining friction and bounce values.

I do agree that the `NAN` default values looks somewhat off though, but I haven't had time yet to look much deeper than that.

tancop commented:
@mihe @Calinou switched to `shape_set_friction_override(RID shape, bool enable, float friction = 0.0)` think its a better solution than two separate methods with more internal state. i agree with mihe rough/absorbent should stay inside `CollisionShape` its combiner specific and might get changed in the future. also fixed setting values on `PhysicsMaterial` at runtime

aaronfranke commented:
It's already been changed but I also want to mention that using NaN is extra awkward since currently the inspector is not able to properly display such values without https://github.com/godotengine/godot/pull/100414 merged in.

tancop commented:
> It's already been changed but I also want to mention that using NaN is extra awkward since currently the inspector is not able to properly display such values without #100414 merged in.

is it ok if `shape_get_friction_override` returns `NaN` when unset or should i return 0 and add another method to check

aaronfranke commented:
@tancop That makes sense to me.

tancop commented:
@mihe @jrouwe you think this is ready to merge with 4.4 out? is there anything i should change

jrouwe commented:
I can only comment about the Jolt integration part and that looks good to me. 

If I can add some minor things (which I think make the code nicer, but I'm fine with the current approach as well):

* You could remove some of the code by moving:

```
void JoltHeightMapShape3D::_update_material(JPH::RefConst<JoltPhysicsMaterial> &p_material) {
	if (!jolt_ref) {
		return;
	}
	jolt_ref_mutex.lock();
	JPH::PhysicsMaterialRefC array[] = { static_cast<JPH::PhysicsMaterialRefC>(p_material) };
	jolt_ref->RestoreMaterialState(array, 1);
	jolt_ref_mutex.unlock();
}
```
to `JoltShape3D` as every shape has a `RestoreMaterialState` call. This allows removing the function from all derived classes. 

* `_get_material` always allocates a new material and if I understand the code correctly, this now gets called unconditionally for every shape on shape construction. Essentially this duplicates the `friction` and `bounce` variables from `JoltShape3D` so you could also opt to store those values only in `JoltPhysicsMaterial` and add a `Ref<JoltPhysicsMaterial>` to `JoltShape3D`. This means you can remove `_update_material` completely as the material will be set on shape construction and when `friction` and `bounce` need to be updated you just update `friction` and `bounce` inside `JoltPhysicsMaterial`. This also allows removing some ifs in `JoltContactListener3D::_override_contact_properties` as nothing is using `JPH::PhysicsMaterial::sDefault`. B.t.w. the concept behind `JPH::PhysicsMaterial::sDefault` is that you can replace it with your own default material, so it is also possible to assign a `JoltPhysicsMaterial` to it with NaN `friction` and `bounce`.

* Having a single material on mesh/height field shapes is quite restrictive in my experience. We allow painting physics materials per triangle (with a max of X materials per mesh). This is maybe something that could be added in the future.

tancop commented:
@jrouwe thanks! just did all that its a lot cleaner that way

tancop commented on modules/jolt_physics/shapes/jolt_shape_3d.h:
```diff
@@ -43,9 +44,11 @@ class JoltShape3D {
+	bool _uses_shape_material = false;
 
-	virtual JPH::ShapeRefC _build() const = 0;
+	virtual JPH::Ref<JPH::Shape> _build() const = 0;
```
just did that :)

jrouwe reviewed (approved):
Looks much better like this!

tancop commented:
@mihe github shows you made a review requesting changes. can you take a look at it again and see if i fixed it? thanks

tancop commented:
> @mihe github shows you made a review requesting changes. can you take a look at it again and see if i fixed it? thanks

happened again after the merge conflict. looks like its a bug with github

tancop commented:
@Calinou anything for the godot physics part?

Calinou reviewed (approved):
Code looks good to me.

AThousandShips commented on modules/godot_physics_2d/godot_body_pair_2d.cpp:
```diff
@@ -239,12 +239,40 @@ bool GodotBodyPair2D::_test_ccd(real_t p_step, GodotBody2D *p_A, int p_shape_A,
+		frictionB = B->get_friction();
+	}
+
+	return ABS(MIN(frictionA, frictionB));
```
```suggestion
	return Math::abs(MIN(frictionA, frictionB));
```

AThousandShips commented on modules/godot_physics_2d/godot_body_pair_2d.cpp:
```diff
@@ -239,12 +239,40 @@ bool GodotBodyPair2D::_test_ccd(real_t p_step, GodotBody2D *p_A, int p_shape_A,
+
+	const GodotShape2D *shapeA = A->get_shape(p_shapeA);
+	frictionA = shapeA->get_friction();
+	if (isnan(frictionA)) {
```
```suggestion
	if (Math::is_nan(frictionA)) {
```

tancop commented on modules/jolt_physics/spaces/jolt_contact_listener_3d.cpp:
```diff
@@ -35,13 +35,20 @@
 #include "../objects/jolt_body_3d.h"
 #include "../objects/jolt_soft_body_3d.h"
+#include "../shapes/jolt_physics_material.h"
+#include "core/error/error_macros.h"
```
left over from testing...

Mickeon commented on doc/classes/CollisionShape2D.xml:
```diff
@@ -27,6 +27,10 @@
 			The margin used for one-way collision (in pixels). Higher values will make the shape thicker, and work better for colliders that enter the shape at a high velocity.
 		</member>
+		<member name="physics_material" type="PhysicsMaterial" setter="set_physics_material" getter="get_physics_material">
+			The shape's physics material.
+			Setting a physics material lets the body have different parts with different properties, like the head and handle on a hammer.
```
No need to be on two different lines. The built-in class reference actually adds another new line in between, so in practice it uses more space than needed.
```suggestion
			The shape's physics material. Setting a physics material lets the body have different parts with different properties, like the head and handle on a hammer.
```

Mickeon commented on doc/classes/CollisionShape3D.xml:
```diff
@@ -39,6 +39,10 @@
 			A disabled collision shape has no effect in the world.
 		</member>
+		<member name="physics_material" type="PhysicsMaterial" setter="set_physics_material" getter="get_physics_material">
+			The shape's physics material.
+			Setting a physics material lets the body have different parts with different properties, like the head and handle on a hammer.
```
```suggestion
			The shape's physics material. Setting a physics material lets the body have different parts with different properties, like the head and handle on a hammer.
```

Mickeon commented on doc/classes/PhysicsServer2D.xml:
```diff
@@ -913,20 +913,43 @@
+			<param index="1" name="enable" type="bool" />
+			<param index="2" name="bounce" type="float" default="0.0" />
+			<description>
+				Sets bounciness for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param bounce] is equivalent to turning on [member PhysicsMaterial.absorbent].
```
Assuming I understand this correctly
```suggestion
				Sets the bounciness for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param bounce] is equivalent to enabling [member PhysicsMaterial.absorbent].
```

tancop replied:
yeah you got that right

Mickeon commented on doc/classes/PhysicsServer2D.xml:
```diff
@@ -944,6 +967,15 @@
+			<param index="1" name="enable" type="bool" />
+			<param index="2" name="friction" type="float" default="0.0" />
+			<description>
+				Sets friction for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param friction] is equivalent to turning on [member PhysicsMaterial.rough].
```
```suggestion
				Sets the friction for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param friction] is equivalent to enabling [member PhysicsMaterial.rough].
```

Mickeon commented on doc/classes/PhysicsServer3D.xml:
```diff
@@ -1016,6 +1030,15 @@
+			<param index="1" name="enable" type="bool" />
+			<param index="2" name="bounce" type="float" default="0.0" />
+			<description>
+				Sets bounciness for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param bounce] is equivalent to turning on [member PhysicsMaterial.absorbent].
```
```suggestion
				Sets the bounciness for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param bounce] is equivalent to enabling [member PhysicsMaterial.absorbent].
```

Mickeon commented on doc/classes/PhysicsServer3D.xml:
```diff
@@ -1024,6 +1047,15 @@
+			<param index="1" name="enable" type="bool" />
+			<param index="2" name="friction" type="float" default="0.0" />
+			<description>
+				Sets friction for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param friction] is equivalent to turning on [member PhysicsMaterial.rough].
```
```suggestion
				Sets the friction for the shape if [param enable] is [code]true[/code], resets to the body's value if [code]false[/code]. A negative value of [param friction] is equivalent to enabling [member PhysicsMaterial.rough].
```

tancop commented:
@Mickeon @AThousandShips anything else?

tancop commented:
@Calinou fixed code style issues the other reviewers brought up

aaronfranke commented on doc/classes/CollisionShape3D.xml:
```diff
@@ -39,6 +39,9 @@
 			A disabled collision shape has no effect in the world.
 		</member>
+		<member name="physics_material" type="PhysicsMaterial" setter="set_physics_material" getter="get_physics_material">
+			The shape's physics material. Setting a physics material lets the body have different parts with different properties, like the head and handle on a hammer.
```
The documentation should explicitly state what happens when both the body and shape have materials. Does the shape's material take priority over the body?

```suggestion
			The shape's physics material. Setting a physics material lets the body have different parts with different properties, such as the head and handle of a hammer.
```

aaronfranke commented on modules/godot_physics_3d/godot_body_pair_3d.cpp:
```diff
@@ -250,12 +250,40 @@ bool GodotBodyPair3D::_test_ccd(real_t p_step, GodotBody3D *p_A, int p_shape_A,
 
-real_t combine_bounce(GodotBody3D *A, GodotBody3D *B) {
-	return CLAMP(A->get_bounce() + B->get_bounce(), 0, 1);
+real_t combine_bounce(GodotBody3D *A, GodotBody3D *B, int p_shapeA, int p_shapeB) {
```
I know this is pre-existing code, but we may want to change this to follow our coding style.

```suggestion
real_t combine_bounce(GodotBody3D *p_body_a, GodotBody3D *p_body_b, int p_shape_a, int p_shape_b) {
```

aaronfranke commented on scene/3d/physics/collision_shape_3d.cpp:
```diff
@@ -281,6 +308,22 @@ bool CollisionShape3D::get_debug_fill_enabled() const {
+		return;
+	}
+
+	RID shape_rid = shape->get_rid();
+
+	if (physics_material.is_null()) {
+		PhysicsServer3D::get_singleton()->shape_set_friction_override(shape_rid, false);
+		PhysicsServer3D::get_singleton()->shape_set_bounce_override(shape_rid, false);
+	} else {
+		PhysicsServer3D::get_singleton()->shape_set_friction_override(shape_rid, true, physics_material->computed_friction());
+		PhysicsServer3D::get_singleton()->shape_set_bounce_override(shape_rid, true, physics_material->computed_bounce());
```
What happens if 2 CollisionShape3D nodes use the same Shape3D resource but have different PhysicsMaterials?

tancop replied:
i dont think theres a way around this without making a full copy of the Shape3D every time we change material. looked at the physics code and material is bound to the shape resource (`JPH::Shape` and `GodotShape3D`) in both engines. we would need to rewrite a lot of godot physics and jolt to change that. i can add some docs to explain this

aaronfranke replied:
I know I suggested before that the CollisionShape3D node should have the physics material, but this only makes sense if this behavior actually happens under the hood. It's really bad if the UI allows this construct but it's broken in the implementation. We either need to change the implementation to support this (preferred, CC @mihe) or go back to how this PR was before with it on the Shape3D resource.

AThousandShips commented on doc/classes/PhysicsServer3DExtension.xml:
These should be documented like the 2D cases, even though the rest of this file is missing descriptions

tancop replied:
doing this for all the methods looks like follow up pr material im def opening one

AThousandShips commented on doc/classes/CollisionShape2D.xml:
```diff
@@ -27,6 +27,9 @@
 			The margin used for one-way collision (in pixels). Higher values will make the shape thicker, and work better for colliders that enter the shape at a high velocity.
 		</member>
+		<member name="physics_material" type="PhysicsMaterial" setter="set_physics_material" getter="get_physics_material">
+		The shape's physics material. Setting a physics material lets the body have different parts with different properties, such as the head and handle of a hammer. Physics materials set on a shape take priority over those set on its body.
```
```suggestion
			The shape's physics material. Setting a physics material lets the body have different parts with different properties, such as the head and handle of a hammer. Physics materials set on a shape take priority over those set on its body.
```

AThousandShips reviewed (approved):
Style and docs look good to me otherwise

AThousandShips commented on doc/classes/PhysicsServer3DExtension.xml:
```diff
@@ -922,6 +922,12 @@
+		<method name="_shape_get_bounce_override" qualifiers="virtual const">
+			<return type="float" />
+			<param index="0" name="shape" type="RID" />
+			<description>
```
You missed this

AThousandShips commented on doc/classes/PhysicsServer3DExtension.xml:
```diff
@@ -934,6 +940,12 @@
+		<method name="_shape_get_friction_override" qualifiers="virtual const">
+			<return type="float" />
+			<param index="0" name="shape" type="RID" />
+			<description>
```
And this

tancop commented:
@Calinou just rebased, you think its ready for merge?

Calinou commented:
@mihe Can you take a final look at this PR? It's good on my end.

mihe reviewed (changes requested):
Apologies for dragging this out. I want to sit down and review this PR in-depth, but making time for lengthy Jolt-related tasks has been tough these past few weeks.

Just quickly trying this out with Jolt I ran into a pretty fundamental issue right away, as seen in my review comment, so that's something that needs to be addressed at least.

I'll set aside time this friday to do a more in-depth review.

tancop commented on modules/jolt_physics/spaces/jolt_contact_listener_3d.cpp:
```diff
@@ -311,6 +326,40 @@ bool JoltContactListener3D::_try_remove_area_overlap(const JPH::SubShapeIDPair &
+	const JoltBody3D *body2 = reinterpret_cast<JoltBody3D *>(p_jolt_body2.GetUserData());
+
+	const uint32_t id1 = p_manifold.mSubShapeID1.GetValue();
+	const uint32_t id2 = p_manifold.mSubShapeID2.GetValue();
```
thanks! i came up with the fix at the same time as you and it works with no error spam

mihe reviewed (changes requested):
I've had a closer look at this PR now, and I have a couple of changes lined up that I'd like to see done for the Jolt side of things.

Instead of going back-and-forth with review comments I decided to just go ahead and make the changes myself, which you'll find in [this diff](https://github.com/tancop/godot/compare/9de1fb42cd2f3c0aa7ce232f2a463da1cb28bbcb...mihe:godot:791f0cf3cbe7991daa5d770d5e75281eb21ebe01).

I've tried to isolate the commits as best as possible, so you should hopefully be able to inspect them individually.

These are the more notable changes I guess:

1. I moved the all the shape material stuff from `JoltShapedObject3D` to `JoltBody3D`. There wasn't much of a reason to have it be shared with `JoltArea3D`, and this maps better to them being stored in `GodotBody3D` for Godot Physics.
1. The updating of manifold reduction was not being done correctly, and resulted in areas enabling it, which they shouldn't do. I also couldn't quite understand why it was being done in `JoltShapedObject3D::commit_shapes`, so that's all similarly been moved to `JoltBody3D`, and now follows the same `thing_changed()` `update_thing()` pattern that's used elsewhere in that file and the module in general, to (hopefully) ensure that this state doesn't go stale.
1. Having `uses_shape_materials()` potentially loop over hundreds of friction values and hundreds of bounce values for complex compound shapes had me concerned, given that it was being called for every contact in the contact listener. I went ahead and added caching of that value instead, which gets updated whenever the shape materials change in any way.
1. Having `_try_override_collision_response` be involved in deciding whether to call `_override_contact_properties` seemed unnecessary. That now instead follows the `_try_do_thing` pattern that's used for the other stuff in the contact listener.
1. I moved the combiner functions into `jolt_math_funcs.h`, to avoid duplicating them in `jolt_contact_listener_3d.cpp`. They probably don't belong there per se, but I couldn't justify creating a whole new file for them.
1. There wasn't much point to storing friction/bounce as `real_t`, since Jolt itself doesn't use double-precision for these, so these now get converted to/from `float` at the physics server boundary, same as a lot of other floating-point values in that module. This technically means you won't always return exactly the same value from the getter that's passed to the setter, but I feel that's a minor concession.

It's up to you how you'd prefer to merge these in, assuming you don't have any objections to the changes, or spot some potential problems with them. I think I have the necessary permissions to push to your branch, if you're fine with that. If not, you should be able to `git apply` [this patch](https://github.com/tancop/godot/compare/9de1fb42cd2f3c0aa7ce232f2a463da1cb28bbcb...mihe:godot:791f0cf3cbe7991daa5d770d5e75281eb21ebe01.patch) as well, or just add my fork as a remote and squash them in.

I haven't looked too closely at the Godot Physics side of the PR, so I can't say much about that, but I do have a couple of concerns, which I've left here as review comments.

tancop committed:
Add support for per-shape physics materials

This lets users override bounce and friction for individual collision shapes instead of using
the same properties for the whole body.

tancop commented on scene/2d/physics/collision_shape_2d.cpp:
```diff
@@ -151,6 +174,8 @@ void CollisionShape2D::set_shape(const Ref<Shape2D> &p_shape) {
 	if (shape.is_valid()) {
 		shape->connect_changed(callable_mp(this, &CollisionShape2D::_shape_changed));
+		// Set material on the new shape
+		_material_changed();
```
yeah thats left over from before i started calling it in `_update_in_shape_owner`

tancop commented on scene/2d/physics/collision_shape_2d.cpp:
```diff
@@ -245,6 +270,18 @@ Color CollisionShape2D::get_debug_color() const {
+	physics_material->connect_changed(callable_mp(this, &CollisionShape2D::_material_changed));
+
+	_material_changed();
+}
```
i decided to write it from scratch instead of copy pasting from 3D and forgot null checks...

tancop commented on scene/2d/physics/collision_shape_2d.cpp:
```diff
@@ -41,6 +41,7 @@ void CollisionShape2D::_shape_changed() {
 
 void CollisionShape2D::_update_in_shape_owner(bool p_xform_only) {
 	collision_object->shape_owner_set_transform(owner_id, get_transform());
+	_material_changed();
```
i wasnt really sure what the function does and played it safe. thanks for the heads up

tancop commented:
@mihe thanks for the review just merged all the changes. I see you removed some casts to `uint32_t` in new `JoltBody3D` methods, had to add them back because `-Wsign-compare` is turned on for official builds

tancop commented:
@mihe this good to go now or you still got questions? just wanna make sure everything clean before it gets merged

mihe reviewed (changes requested):
I found a fairly critical issue with the Godot Physics implementation, and some very minor cosmetic remarks in the Jolt implementation.

Outside of these review comments, I must admit that I'm becoming more and more uncomfortable with the API of these two `PhysicsServer*D` methods:

```cpp
void body_set_shape_friction_override(RID p_body, int p_shape_idx, bool p_enable, real_t p_friction = 0.0);
void body_set_shape_bounce_override(RID p_body, int p_shape_idx, bool p_enable, real_t p_bounce = 0.0);
```

I realize it's a bit late to be having this conversation, and that the current design stems from feedback that have been given here already, so I won't push too hard on this, but the whole thing with having `p_enable` there, and then defaulting the values to `0.0`, just so you can pass `false` without needing to give it a value. It makes for an odd-looking API in my opinion.

Unfortunately I struggle to think of an alternative that wouldn't ideally require a `bool` to be added for each material state as well, which seems like a hefty price to pay for what is only (subjectively) a better API.

Anyway, I just figured I'd voice my concern at least, in case anyone had any other ideas, or know of other Godot APIs that do something similar already.

mihe commented on scene/3d/physics/collision_shape_3d.h:
```diff
@@ -32,12 +32,14 @@
 	GDCLASS(CollisionShape3D, Node3D);
 
 	Ref<Shape3D> shape;
+	Ref<PhysicsMaterial> physics_material;
```
Looking at the relationship between `CollisionShape*D` and `CollisionObject*D` more closely, it would seem to me that this is not quite correct, and should instead be implemented much in the same way that the `disabled` property is currently implemented, where it should live in `CollisionObject3D::ShapeData` instead, and there should be `shape_owner_*` methods (that are exposed to scripting) like `shape_owner_set_physics_material` and such.

I frankly don't quite understand the purpose of all this "shape owner" stuff, and why there isn't just one shape owner per object, but again, this new physics material sits alongside the `disabled` property in `CollisionShape*D`, so there's no reason why it shouldn't be implemented in the same way.

Calinou replied:
We have several properties that do nothing on certain subclasses (Light3D is an example of this with some of its properties only being usable with OmniLight3D and SpotLight3D), so I guess it's fine to do this way.

tancop commented:
@mihe switched this to use shape owner methods can you take a look if i did it right

tancop commented:
@Calinou looks like mihe doesnt have time to review. is anyone else free right now?

Calinou reviewed (approved):
Tested again, it works as expected with GodotPhysics3D and Jolt Physics. Code and documentation look good to me.

I suggest having mihe take a final look at the API still, just in case. We can't change what's exposed once it lands in a stable release, so we need to make sure it's rock-solid.

mihe commented:
I'll try to find some time this week to do another pass on this.

> We can't change what's exposed once it lands in a stable release, so we need to make sure it's rock-solid.

I will say that I am leaning towards suggesting that we do go with this API for `PhysicsServer*D` after all:

```gdscript
body_set_shape_friction_override(body: RID, shape_idx: int, friction: float) -> void
body_set_shape_bounce_override(body: RID, shape_idx: int, bounce: float) -> void

body_get_shape_friction_override(body: RID, shape_idx: int) -> float
body_get_shape_bounce_override(body: RID, shape_idx: int) -> float

body_set_enable_shape_friction_override(body: RID, shape_idx: int, enable: bool) -> void
body_set_enable_shape_bounce_override(body: RID, shape_idx: int, enable: bool) -> void

body_is_shape_friction_override_enabled(body: RID, shape_idx: int) -> bool
body_is_shape_bounce_override_enabled(body: RID, shape_idx: int) -> bool
```

However, what I'm still unsure of is whether or not we would mandate that enabling/disabling should be completely disconnected from the value itself or not, where the value would persist even after disabling.

If the value must in fact persist even after disabling, then that of course changes how we need to store things, but we could also just document that:

1. Setting the value implicitly enables the override.
2. Disabling the override effectively resets the value.

... which would let implementers (i.e. our physics servers and any third-party extensions) omit any sort of bool/bit/flag for the enabled state, but I'm not sure if we're just trading in one awkward API for another in that case.

Also, if all we're doing is trying to save on memory from this enabled bool/bit/flag, then why are we going with the "dense" storage (i.e. `LocalVector<float>`) for these material values in the first place? Wouldn't the more common use-case of per-shape materials be that you rely on the body's material for the majority of the shapes, and then selectively set a few shapes to override that material? If so, then you would ideally store this in some kind of "sparse" storage anyway, like a shape-index-to-material-value map, at which point the enabled state comes "for free" by virtue of the shape index' presence in this map.

If you'll allow me to generalize even more, the vast majority of bodies created will likely not use these overrides, and if they do, they will likely consist of only a handful of shapes, at which point the cost/benefit analysis of dense vs sparse storage is likely moot anyway.

So maybe we do in fact keep the material values separate from their enabled state, and thus end up with what I would argue is a (mostly) better API?

mihe reviewed (changes requested):
I did another pass on this and found some bugs with the recent changes, which you'll find in the review comments. I also threw in some more nitpicks while I was at it.

I've also spent some more time thinking about the `PhysicsServer*D` API, and I think I've arrived at a better API for this:

```gdscript
func body_set_shape_friction_override(body: RID, shape_idx: int, friction: float) -> void
func body_get_shape_friction(body: RID, shape_idx: int) -> float
func body_is_shape_friction_overridden(body: RID, shape_idx: int) -> bool
func body_clear_shape_friction_override(body: RID, shape_idx: int) -> void
```

This gets rid of `NAN` from the API completely, by having `body_get_shape_friction` fall back on the body's friction when there is no override set, while still not burdening implementers with a separate enabled state.

To save us some time, I've gone ahead and made these changes already, and just like before you'll find the changes [here](https://github.com/tancop/godot/compare/de20ef008f7c2f0639000582adac0b4584afd5cb...mihe:godot:add-per-shape-materials) (patch [here](https://github.com/tancop/godot/compare/de20ef008f7c2f0639000582adac0b4584afd5cb...mihe:godot:add-per-shape-materials.patch)), which also includes fixes for the new review comments.

Like before, if you have no objections to these changes, I can go ahead and push directly to your fork/branch as well. We're only a few days away from the 4.5 feature freeze, and I don't see anything else in this PR that should prevent this from being merged by then.

EDIT: Changes rebased on top of latest push (de20ef008f7c2f0639000582adac0b4584afd5cb).

mihe commented:
As another last-minute decision, I've omitted the `body_is_shape_(friction|bounce)_overridden` physics server methods from the API that I suggested above, because I struggle to see a use-case for such a query, and it's not something that's needed by the scene nodes anyway.

If anyone needs this in the future it can trivially be added then instead.

So the `PhysicsServer*D` API would look like this instead, which I'm mostly happy with:

```gdscript
func body_set_shape_friction_override(body: RID, shape_idx: int, friction: float) -> void
func body_clear_shape_friction_override(body: RID, shape_idx: int) -> void
func body_get_shape_friction(body: RID, shape_idx: int) -> float
```

The branch/patch linked above should have updated accordingly, but [here](https://github.com/tancop/godot/compare/269885f36d4af7365d539140986acecb366c4ecb...mihe:godot:add-per-shape-materials) (patch [here](https://github.com/tancop/godot/compare/269885f36d4af7365d539140986acecb366c4ecb...mihe:godot:add-per-shape-materials.patch)) it is again if you need it.

Xtarsia commented:
Does this have any chance of working with Heightmap Shapes?

In Terrain3D I would like to be able to send a physics material "ID" for each Quad. Since We have access to an index map that could determine the dominant texture for any given quad, and that texture has other properties associated with it already, linking each texture ID (upto 32) to a physics material, would essentially Let us have paintable physics materials.

mihe commented:
> Does this have any chance of working with Heightmap Shapes? [...] In Terrain3D I would like to be able to send a physics material "ID" for each Quad.

@Xtarsia Not as-is, no. The same goes for individual triangles in `ConcavePolygonShape3D`. It's technically doable with Jolt, which supports this for both shapes, but it's not immediately clear to me how you'd go about implementing it for Godot Physics.

Either way, we'd need some way of actually assigning material values to groups of faces, both from a UX perspective and from an API perspective, which is likely quite a bit of work, and better suited for a future PR. The texture approach might make some sense for `HeightMapShape3D`, but I'm not sure it suits `ConcavePolygonShape3D` as well.
